### PR TITLE
Reducing `picking.py` to be simpler, more focused, and more performant.

### PR DESCRIPTION
**Pull request by tfenne**
Closes https://github.com/fulcrumgenomics/prymer/issues/38

### CODE REVIEW

**Comment by tfenne** [comment]
I'm still actively working on this - mostly writing tests and fixing up documentation.

File: prymer/api/picking.py [hunk 1]
```diff
@@ -186,285 +79,108 @@ def score(
...
-    primer_pairs: list[PrimerPair] = []
-    for left_primer in left_primers:
-        for right_primer in right_primers:
-            if left_primer.span.refname != right_primer.span.refname:
-                raise ValueError(
-                    "Cannot create a primer pair from left and right primers on different"
-                    f" references: left: '{left_primer.span.refname}'"
-                    f" right: '{right_primer.span.refname}'"
+    # Short circuit if we have no left primers or no right primers
+    if not any(left_primers) or not any(right_primers):
```
Comment (msto) [inline prymer/api/picking.py:134]: **issue** If either `left_primers` or `right_primers` is a generator, `any()` will consume an element.

**suggestion** Consider converting the primers to a `list`, or find a solution to peek at the first element without consuming it.

File: prymer/api/picking.py [hunk 2]
```diff
@@ -186,285 +79,108 @@ def score(
...
+        # generate all the primer pairs that don't violate hard size and Tm constraints
+        for lp in left_primers:
+            for rp in right_primers:
+                if rp.span.end - lp.span.start + 1 > amplicon_sizes.max:
+                    continue
+
+                amp_mapping = Span(refname=target.refname, start=lp.span.start, end=rp.span.end)
+                amp_bases = bases[
+                    amp_mapping.start - region_start : amp_mapping.end - region_start + 1
+                ]
```
Comment (msto) [inline prymer/api/picking.py:158]: **question** Is this replicating functionality that is already available via `PrimerPair`? I'm surprised that amplicon length, span, and bases wouldn't be available through that API.

File: prymer/api/picking.py [hunk 3]
```diff
@@ -186,285 +79,108 @@ def score(
...
+                amp_tm = calculate_long_seq_tm(amp_bases)
```
Comment (msto) [inline prymer/api/picking.py:159]: **suggestion** Related to the last comment, should this be a `cached_property` on `PrimerPair`?

File: prymer/api/picking.py [hunk 4]
```diff
@@ -186,285 +79,108 @@ def score(
...
+
+                if amp_tm < amplicon_tms.min or amp_tm > amplicon_tms.max:
+                    continue
+
+                if max_heterodimer_tm is not None:
+                    if ntthal.duplex_tm(lp.bases, rp.bases) > max_heterodimer_tm:
+                        continue
+
+                penalty = score(
```
Comment (msto) [inline prymer/api/picking.py:168]: **question** Naive question - why is `duplex_tm()` a method on `NtThermoAligner`, but `score()` and `calculate_long_seq_tm()` are standalone functions? Are the latter accessible through `primer3`?

**Review by msto — CHANGES_REQUESTED** [review]

File: prymer/api/picking.py (hunk 1 above)
Comment (tfenne) [reply prymer/api/picking.py:134]: Ugh, I always forget the `Iterator` in python is `Iterable`.

File: prymer/api/picking.py (hunk 4 above)
Comment (tfenne) [reply prymer/api/picking.py:168]: `duplex_tm()` is a complicated thing that does a thermodynamic alignment of two sequences and predicts their Tm, which is why we delegate to ntthal instead of trying to replicate that.

`score()` is mimicking something in primer3, but is not easily accessible independent of designing primer pairs in primer3.

`calculate_long_seq_tm()` is so simple it's not worth trying to redirect to primer3.

File: prymer/api/picking.py [hunk 5]
```diff
@@ -186,285 +79,109 @@ def score(
...
+    # Grab the sequence we'll use to fill in the amplicon sequence
+    with FastaFile(f"{fasta_path}") as fasta:
+        region_start = min(p.span.start for p in left_primers)
+        region_end = max(p.span.end for p in right_primers)
+        bases = fasta.fetch(target.refname, region_start - 1, region_end)
+
+    with NtThermoAlign() as ntthal:
+        # generate all the primer pairs that don't violate hard size and Tm constraints
+        for lp in left_primers:
+            for rp in right_primers:
```
Comment (clintval) [inline prymer/api/picking.py:152]: `itertools.product` would help you un-nest this a level:

https://docs.python.org/3/library/itertools.html#itertools.product

File: prymer/api/picking.py (hunk 2 above)
Comment (tfenne) [reply prymer/api/picking.py:159]: Not really ... PrimerPair does have a function to make the amplicon mapping from the two primers, but it's trivial and also it's not a static/class method in PrimerPair :/

PrimerPair does not have any functionality to fetch the bases for the amplicon - it expected to be given it.

File: prymer/api/picking.py (hunk 3 above)
Comment (tfenne) [reply prymer/api/picking.py:160]: I'd prefer not ... we have on implementation of long sequence Tm, but it's possible other versions might be used.

File: prymer/api/picking.py [hunk 6]
```diff
@@ -186,285 +69,112 @@ def score(
...
+    amplicon_tms: MinOptMax[float],
+    max_heterodimer_tm: Optional[float],
+    weights: PrimerAndAmpliconWeights,
+    fasta_path: Path,
+) -> Iterator[PrimerPair]:
     """Builds primer pairs from individual left and primers.
 
+    Only primer pairs that meet the requirements for amplicon sizes and tms will be returned.
+    In addition, if `max_heterodimer_tm` is provided then primer pairs with a heterodimer tm
+    exceeding the maximum will also be discarded.
```
Comment (msto) [inline prymer/api/picking.py:107]: **question** Why is this necessary when `primer3` will apply these constraints to designed primer pairs if they are specified? e.g. specifying `PRIMER_PRODUCT_MIN_TM` and `PRIMER_PRODUCT_MAX_TM` will constrain the pairs returned by `primer3` to the provided range

**Review by msto — COMMENTED** [review]
I don't have any objections to the refactoring.

I do continue to struggle to understand the motivation for reimplementing primer pair picking and scoring in a less performant language, rather than using `primer3` to design and filter pairs where possible.

I understand that `primer3` does not support all the possible flavors of filtering we sometimes have to implement (e.g. I don't think its design algorithm supports filtering based on heterodimerization tm), but I think that it's generally simpler and less prone to inconsistency to rely on it for the features it does provide.

File: prymer/api/picking.py (hunk 6 above)
Comment (tfenne) [reply prymer/api/picking.py:107]: Because this is used when designing left and right primers independently - so primer3 is not emitting pairs and not cannot check these constraints.
